UNITTESTS/drivers/PwmOut/test_pwmout.cpp

```diff
@@ -0,0 +1,202 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+#include "drivers/PwmOut.h"
+
+using namespace mbed;
+
+class MbedPowerMgmtInterface {
+public:
+    virtual ~MbedPowerMgmtInterface() {}
+    virtual void sleep_manager_unlock_deep_sleep_internal() = 0;
+    virtual void sleep_manager_lock_deep_sleep_internal() = 0;
+};
+
+
+class MockMbedPowerMgmt : public MbedPowerMgmtInterface {
+public:
+    virtual ~MockMbedPowerMgmt() {}
+    MOCK_METHOD0(sleep_manager_unlock_deep_sleep_internal, void());
+    MOCK_METHOD0(sleep_manager_lock_deep_sleep_internal, void());
+
+    static MockMbedPowerMgmt *get_instance(bool create_instance)
+    {
+        if (instance == nullptr && create_instance) {
+            instance = new MockMbedPowerMgmt();
+        }
+
+        if (create_instance) {
+            ++ref_counter;
+        }
+
+        return instance;
+    }
+
+    static void release_instance()
+    {
+        --ref_counter;
+
+        if ((instance != nullptr) && (ref_counter == 0)) {
+            delete instance;
+            instance = nullptr;
+        }
+    }
+
+    MockMbedPowerMgmt(MockMbedPowerMgmt const&) = delete;
+    void operator=(MockMbedPowerMgmt const&) = delete;
+private:
+    MockMbedPowerMgmt() {}
+
+    static MockMbedPowerMgmt *instance;
+    static int ref_counter;
+};
+MockMbedPowerMgmt *MockMbedPowerMgmt::instance = nullptr;
+int MockMbedPowerMgmt::ref_counter = 0;
+
+
+void sleep_manager_unlock_deep_sleep(void)
+{
+    // Get the singleton if a test case has created one.
+    // We do not want to increment the call count unless it's done by a test
+    // case.
+    MockMbedPowerMgmt *mbed_mgmt_mock = MockMbedPowerMgmt::get_instance(false);
+    if (mbed_mgmt_mock != nullptr) {
+        // Call only if the singleton was created by a test case.
+        mbed_mgmt_mock->sleep_manager_unlock_deep_sleep_internal();
+    }
+}
+
+
+void sleep_manager_lock_deep_sleep(void)
+{
+    // Get the singleton if a test case has created one.
+    // We do not want to increment the call count unless it's done by a test
+    // case.
+    MockMbedPowerMgmt *mbed_mgmt_mock = MockMbedPowerMgmt::get_instance(false);
+    if (mbed_mgmt_mock != nullptr) {
+        // Call only if the singleton  was created by a test case.
+        mbed_mgmt_mock->sleep_manager_lock_deep_sleep_internal();
+    }
+}
+
+
+class TestPwmOut : public testing::Test {
+protected:
+    PwmOut *pwm_obj;
+    MockMbedPowerMgmt *mbed_mgmt_mock;
+
+    void SetUp()
+    {
+        pwm_obj = new PwmOut(PTC1);
+        // Create a mock object singleton after the PwmOut object
+        // instantiation so the sleep_manager_lock_deep_sleep call by the
+        // constructor doesnot increment the call count.
+        // Now it is ok because a test case is about to start.
```

Comment:
But we should also test that at creation the PwmOut object calls sleep_manager_lock_deep_sleep.
Can you check at the beginning of the test case that the deep_sleep lock is held and then run your test case?

Reply:
>@jamesbeyond suggested that we manually run the FPGA test shield tests which test pwm.
In addition we should add tests for the new suspend/resume API in https://github.com/ARMmbed/mbed-os/blob/master/TESTS/mbed_hal_fpga_ci_test_shield/pwm/main.cpp

I added test for `resume` and `suspend` not for the constructor.

Reply:
But you are adding a lot of complexity in order not to test something that adds value...

Reply:
If we want to test the constructor a separate test case should be added. 

The added test cases are for the two methods I mentioned. The expectations in the test cases are clear without having to explain anything with a comment (`// Add an extra expected call because the first call is by the constructor`). 
What if the constructor changes and no longer calls the deep sleep method in the future? We will have to come and change the expectation number again for the `resume` and `suspend` test cases even if those two methods have not been changed. We will have a dependency on the constructor. This way it is fixed as it only check the call count by `resume`/`suspend.`